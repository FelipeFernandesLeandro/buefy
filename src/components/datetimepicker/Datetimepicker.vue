<template>
    <b-datepicker
        v-if="!isMobile || inline"
        ref="datepicker"
        v-model="computedValue"
        v-bind="datepicker"
        :open-on-focus="openOnFocus"
        :inline="inline"
        :editable="editable"
        :close-on-click="false"
        :date-formatter="defaultDatetimeFormatter"
        :date-parser="defaultDatetimeParser">
        <nav class="level is-mobile">
            <div
                class="level-item has-text-centered"
                v-if="$slots.left !== undefined">
                <slot name="left" />
            </div>
            <div class="level-item has-text-centered">
                <b-timepicker
                    ref="timepicker"
                    v-bind="timepicker"
                    v-model="computedValue"
                    inline
                    :editable="editable"
                />
            </div>
            <div
                class="level-item has-text-centered"
                v-if="$slots.right !== undefined">
                <slot name="right" />
            </div>
        </nav>
    </b-datepicker>
    <b-input
        v-else
        ref="input"
        type="datetime-local"
        autocomplete="off"
        :value="formatNative(computedValue)"
        :placeholder="placeholder"
        :size="size"
        :icon="icon"
        :icon-pack="iconPack"
        :loading="loading"
        :max="formatNative(maxDate)"
        :min="formatNative(minDate)"
        :disabled="disabled"
        :readonly="false"
        v-bind="$attrs"
        :use-html5-validation="useHtml5Validation"
        @change.native="onChangeNativePicker"
        @focus="onFocus"
        @blur="onBlur"/>
</template>

<script>
import FormElementMixin from '../../utils/FormElementMixin'
import { isMobile } from '../../utils/helpers'
import config from '../../utils/config'

import Datepicker from '../datepicker/Datepicker'
import Timepicker from '../timepicker/Timepicker'

export default {
    name: 'BDatetimepicker',
    components: {
        [Datepicker.name]: Datepicker,
        [Timepicker.name]: Timepicker
    },
    mixins: [FormElementMixin],
    inheritAttrs: false,
    props: {
        value: {
            type: Date
        },
        editable: {
            type: Boolean,
            default: false
        },
        inline: Boolean,
        openOnFocus: Boolean,
        mobileNative: {
            type: Boolean,
            default: true
        },
        datetimeFormatter: {
            type: Function
        },
        datetimeParser: {
            type: Function
        },
        timeCreator: {
            type: Function,
            default: (date) => {
                if (typeof config.defaultDatetimeTimeCreator === 'function') {
                    return config.defaultDatetimeTimeCreator(date)
                } else {
                    return date
                }
            }
        },
        datepicker: Object,
        timepicker: Object
    },
    data() {
        return {
            newValue: this.value
        }
    },
    computed: {
        computedValue: {
            get() {
                return this.newValue
            },
            set(value) {
                let val = value
                if (value !== null && typeof value !== 'undefined') {
                    val = new Date(value.getTime())
                    if (this.newValue) {
                        // on datepicker select -> maintain time part
                        if (value.getHours() === 0 &&
                            value.getMinutes() === 0 &&
                            value.getSeconds() === 0) {
                            val.setHours(this.newValue.getHours(),
                                this.newValue.getMinutes(),
                                this.newValue.getSeconds(), 0)
                        }
                    } else {
                        val = this.timeCreator(val)
                    }
                }
                this.newValue = val
                this.$emit('input', value)
            }
        },
        isMobile() {
            return this.mobileNative && isMobile.any()
        }
    },
    watch: {
        value(value) {
            this.newValue = value
        }
    },
    methods: {
        defaultDatetimeParser(date) {
            if (typeof this.datetimeParser === 'function') {
                return this.datetimeParser(date)
            } else if (typeof config.defaultDatetimeParser === 'function') {
                return config.defaultDatetimeParser(date)
            } else {
                return new Date(Date.parse(date))
            }
        },
        defaultDatetimeFormatter(date) {
            if (typeof this.datetimeFormatter === 'function') {
                return this.datetimeFormatter(date)
            } else if (typeof config.defaultDatetimeParser === 'function') {
                return config.defaultDatetimeParser(date)
            } else {
                const yyyyMMdd = date.getFullYear() +
                    '/' + (date.getMonth() + 1) +
                    '/' + date.getDate()
                const d = new Date(yyyyMMdd)
                return d.toLocaleDateString() +
                    ' ' +
                    this.$refs.timepicker.timeFormatter(date, this.$refs.timepicker)
            }
        },
        /*
        * Parse date from string
        */
        onChangeNativePicker(event) {
            const date = event.target.value
            this.computedValue = date ? new Date(date) : null
        },
        formatNative(value) {
            const date = new Date(value)
            if (value && !isNaN(date)) {
                const year = date.getFullYear()
                const month = date.getMonth() + 1
                const day = date.getDate()
                const hours = date.getHours()
                const minutes = date.getMinutes()
                const seconds = date.getSeconds()
                return year + '-' +
                    ((month < 10 ? '0' : '') + month) + '-' +
                    ((day < 10 ? '0' : '') + day) + 'T' +
                    ((hours < 10 ? '0' : '') + hours) + ':' +
                    ((minutes < 10 ? '0' : '') + minutes) + ':' +
                    ((seconds < 10 ? '0' : '') + seconds)
            }
            return ''
        }
    }
}
</script>
