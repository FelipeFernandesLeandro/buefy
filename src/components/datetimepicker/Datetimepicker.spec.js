import { shallowMount } from '@vue/test-utils'
import Datetimepicker from '@components/datetimepicker/Datetimepicker'

let wrapper

describe('Datetimepicker', () => {
    beforeEach(() => {
        wrapper = shallowMount(Datetimepicker)
    })

    it('is called', () => {
        expect(wrapper.name()).toBe('BDatetimepicker')
        expect(wrapper.isVueInstance()).toBeTruthy()
    })

    it('render correctly', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
})
