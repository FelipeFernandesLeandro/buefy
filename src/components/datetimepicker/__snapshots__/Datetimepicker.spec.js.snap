// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datetimepicker render correctly 1`] = `
<bdatepicker-stub daynames="Su,M,Tu,W,Th,F,S" monthnames="January,February,March,April,May,June,July,August,September,October,November,December" firstdayofweek="0" dateformatter="function boundFn (a) {
    var l = arguments.length;
    return l
      ? l > 1
        ? fn.apply(ctx, arguments)
        : fn.call(ctx, a)
      : fn.call(ctx)
  }" dateparser="function boundFn (a) {
    var l = arguments.length;
    return l
      ? l > 1
        ? fn.apply(ctx, arguments)
        : fn.call(ctx, a)
      : fn.call(ctx)
  }" datecreator="function _default() {
                    /* istanbul ignore next */cov_1yh4yim62k.f[11]++;
                    cov_1yh4yim62k.s[59]++;

                    if (typeof _config2.default.defaultDateCreator === 'function') {
                        /* istanbul ignore next */cov_1yh4yim62k.b[15][0]++;
                        cov_1yh4yim62k.s[60]++;

                        return _config2.default.defaultDateCreator();
                    } else {
                        /* istanbul ignore next */cov_1yh4yim62k.b[15][1]++;
                        cov_1yh4yim62k.s[61]++;

                        return new Date();
                    }
                }" mobilenative="true" indicators="dots" iconprev="chevron-left" iconnext="chevron-right" yearsrange="-100,3" nearbymonthdays="true" rulesforfirstweek="4">
    <nav class="level is-mobile">
        <!---->
        <div class="level-item has-text-centered">
            <btimepicker-stub inline=""></btimepicker-stub>
        </div>
        <!---->
    </nav>
</bdatepicker-stub>
`;
